multirun/multirun.go

```diff
@@ -10,49 +10,97 @@ type multirun struct {
 	commands               []command
 	stdoutSink, stderrSink io.Writer
 	args                   []string
-	parallel               bool
+	jobs                   int
 	quiet                  bool
 }
 
 func (m multirun) run(ctx context.Context) error {
-	if m.parallel {
-		errs := make(chan error)
-		for _, cmd := range m.commands {
-			p := process{
-				tag:        cmd.Tag,
-				path:       cmd.Path,
-				stdoutSink: m.stdoutSink,
-				stderrSink: m.stderrSink,
-				args:       m.args,
-				addTag:     true,
-			}
-			go func() {
-				errs <- p.run(ctx)
-			}()
+	// minimum concurrency is 1, which is sequential
+	if m.jobs < 0 {
+		return fmt.Errorf("jobs must be at least 0")
+	}
+
+	if m.jobs == 0 {
+		return m.unboundedExecution(ctx)
+	}
+
+	return m.boundedExecution(ctx)
+}
+
+// unboundedExecution execute multiple commands without concurrency limit
+func (m multirun) unboundedExecution(ctx context.Context) error {
+	errs := make(chan error)
+
+	for _, cmd := range m.commands {
+		p := process{
+			tag:        cmd.Tag,
+			path:       cmd.Path,
+			stdoutSink: m.stdoutSink,
+			stderrSink: m.stderrSink,
+			args:       m.args,
+			addTag:     true,
 		}
-		var firstError error
-		for range m.commands {
-			if err := <-errs; firstError == nil {
-				firstError = err
-			}
+
+		go func() {
+			errs <- p.run(ctx)
+		}()
+	}
+
+	var firstError error
+	for range m.commands {
+		if err := <-errs; firstError == nil {
+			firstError = err
 		}
-		return firstError
-	} else {
-		for _, cmd := range m.commands {
-			p := process{
-				tag:        cmd.Tag,
-				path:       cmd.Path,
-				stdoutSink: m.stdoutSink,
-				stderrSink: m.stderrSink,
-				args:       m.args,
-			}
-			if !m.quiet {
-				fmt.Fprintf(m.stderrSink, "Running %s\n", cmd.Tag)
-			}
-			if err := p.run(ctx); err != nil {
-				return err
-			}
+	}
+
+	return firstError
+}
+
+// boundedExecution execute multiple commands using a sized worker pool
+func (m multirun) boundedExecution(ctx context.Context) error {
+	commands := make(chan command)
+	errs := make(chan error)
+
+	// start worker pool
+	for w := 0; w < m.jobs; w++ {
+		go m.spawnWorker(ctx, commands, errs)
+	}
+
+	// send command to worker pool
+	for _, cmd := range m.commands {
+		commands <- cmd
```

Comment:
This will block forever if there are more commands than workers because workers will block on sending to the `errs` channel but nobody is reading. Either `errs` or `commands` needs to be a buffered channel. I suggest the former because it'll take a bit less memory (`error` is smaller than `command`).

Reply:
Good find. Will fix!